### PR TITLE
[Demo] [Incomplete] Allow compilation and linking against BoringSSL

### PR DESCRIPTION
**Pull request by gpshead**
NOTE: This is focused solely on the C code changes required. It is not usable as is. Missing items if you were to attempt to use this yourself: build system changes necessary for this to work have not been completed -- we don't use CPython's build system when building the runtime at Google.

The purpose of opening this as a Draft PR is to show what kinds of changes are _currently_ needed. I'm not aiming to see this merged as is - though we may choose pull bits and pieces out into their own small PRs that could make sense on their own to minimize the changes necessary for those using other TLS libraries where easily maintainable.

It may be of potential interest to some ongoing discussions (I expect to link to it) and issue #116333 authors and its draft PR #116334.

---

Playing around with hackish _quick_ attempts (mostly not yet included in the PR branch) at using CPython's existing build system:
 * BoringSSL libssl.a requires C++ linkage.  The configure check for `_ssl` module being available fails otherwise.
 * The wrong OpenSSL headers appear to get included when compiling `Modules/_hashopenssl.c`
 * You must statically link BoringSSL and thus the `_ssl` and `_hashlib` modules as that is its intent (this requires `Modules/Setup*` changes)

Proper configure.ac, Setup, & Makefile.pre.in plumbing would be needed to alleivate all of this.

----
📚 Documentation preview 📚: https://cpython-previews--116399.org.readthedocs.build/

### CODE REVIEW

**Comment by gpshead** [comment]
FYI for @twouters & @davidben - This serves as a public example of most of the changes needed to link CPython against BoringSSL.

**Comment by davidben** [comment]
For some context, some of the changes can probably be reduced. In particular, I plan to implement the "ex" versions of SSL_read and SSL_write. We'd skipped them originally because OpenSSL messed them up. (They forgot EOF was different from error.) But between a fix I landed in CPython a bit ago and some ideas we had on our end, I think we can implement them without picking up the problems from OpenSSL's version.

Likewise I think the way to resolve errors is to add new API to return the symbol name of things. OpenSSL hasn't been responsive to this feature request but I plan to just go ahead and add them on our end.

File: Modules/_ssl.c [hunk 1]
```diff
@@ -4542,50 +4582,6 @@ set_sni_callback(PySSLContext *self, PyObject *arg, void *c)
     return 0;
 }
 
-#if OPENSSL_VERSION_NUMBER < 0x30300000L
```
Comment (davidben) [inline Modules/_ssl.c:4544]: Oh BoringSSL supports get1_objects. My thinking was we'd just update this ifdef to skip the polyfill in BoringSSL and then leave the rest as-is. get1_objects avoids a race condition in the old CPython code. (get0_objects was added to OpenSSL for CPython, but it turned out to have been the wrong API the whole time. get1_objects was to undo that mistake)

**Comment by awilfox** [comment]
From the [BoringSSL README](https://github.com/google/boringssl#boringssl):

> Although BoringSSL is an open source project, it is not intended for general use, as OpenSSL is. We don't recommend that third parties depend upon it. Doing so is likely to be frustrating because there are no guarantees of API or ABI stability.

It seems odd to then add support for building Python against BoringSSL, unless there will be a guarantee of API/ABI stability.  If that is indeed the case, which version will be considered stable?  That would be the one that CPython should target.  If this lands for 3.14, such a theoretical BoringSSL API/ABI would need to be supported until October 2030; what would the maintenance cycle look like?  If there is a security issue requiring an API change that is relevant to already-released CPython versions, what would the disclosure process look like?

**Comment by gpshead** [comment]
> It seems odd to then add support for building Python against BoringSSL, unless there will be a guarantee of API/ABI stability.

Shipping with it is not the point of this kind of work.  There isn't a support question as nobody doing this ships binaries who does not own and manage their full software stack themselves.  CPython has no plans to ship a binary release using BoringSSL or AWS-LC unless a [PEP-11](https://peps.python.org/pep-0011/) tier 1 or 2 platform that we ship binaries for were based on it.

BoringSSL is what my former employer has been using with its own Python runtime for almost a decade.

It is also what we've had another contributor apparently working for Amazon to push changed to CPython for it to build and work with Amazon's fork of Google's [BoringSSL](https://boringssl.googlesource.com/boringssl) known as [AWS-LC](https://github.com/aws/aws-lc/).  Both libraries get used in very real huge scale production circumstances even if a _typical_ end user doesn't realize it.  Upstreaming the patches to make CPython itself work without needing patches simplifies the life of integrators.

I'm sympathetic to anyone who wants to contribute patches for this kind of thing and can help review PRs to go in, but I'm otherwise dropping doing all such work myself unless needed for an employer.

It'd be great if someone who cared about this would also [contribute a CPython buildbot](https://www.python.org/dev/buildbot/) configured to build against your relevant AWS-LC or BoringSSL library.

**Pull request closed without merging**